src/pipelinedefinitionstatus.hpp

```diff
@@ -0,0 +1,288 @@
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::LOADING_PRECONDITION_FAILED_REQUIRED_REVALIDATION;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct RetiredState {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::RETIRED;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+    StateKeeper handle(const RetireEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+};
+
+class PipelineDefinitionStatus : public MachineState<BeginState, AvailableState, AvailableRequiredRevalidation, LoadingPreconditionFailedState, LoadingFailedLastValidationRequiredRevalidation, RetiredState> {
+public:
+    bool isAvailable() {
+        auto state = getState();
+        return (state == PipelineDefinitionStateCode::AVAILABLE) ||
+               (state == PipelineDefinitionStateCode::AVAILABLE_REQUIRED_REVALIDATION);
+    }
+    bool canEndLoaded() {
```

Comment:
Maybe `canEndAvailable` is better name to keep conventions? We never have "Loaded" state.

Reply:
What is purpose of this method? We can end up with available method even if we have method retired in first place.

Reply:
Its purpose is to check if with current pipeline status it is reasonable to assume that it can be loaded & available soon.
Eg. we fixed the model configuration so that pipeline could load but its previous validation fails.

But that being said this may be an overkill and we probably could not even care for this transitional state.

Reply:
isLoadingOrAvailble can be alternative name

Reply:
The expectation with this function is to check if it is either already loaded(AVAILALBE, AVAILABLE_REQUIRED_REVALIDATION) or not loaded due to failure in previous validation but with information that there will be revalidation shortly since one of used models changed (LOADING_PRECONDITION_FAILED_REQUIRED_REVALIDATION).

Other possible name could be eg 
isLoadedOrLoadingInProgress?

We still want to distinguish this from LOADING_PRECONDITION_FAILED when we know there is nothing to wait for.